tests/common/mock_environment.py

```diff
@@ -0,0 +1,149 @@
+    It can be useful in tests that need to execute commands or access/modify files that are not available in all platforms,
+    or require root privileges, or can change global system settings. For a sample usage see the mock_cgroup_environment()
+    function.
+
+    Currently, MockEnvironment mocks subprocess.Popen(), fileutil.mkdir(), os.path.exists() and the builtin() open function
+    (it mocks fileutil.mkdir() instead od os.mkdir() because the agent's code users the former since it provides
+    backwards compatibility with Python 2).
+
+    The mock for Popen looks for a match in the given 'commands' and, if found, forwards the call to the the echo command
+    to produce the output for the matching item. Otherwise it forwards the call to the original Popen function.
+
+    The mocks for the other functions first look for a match in the given 'files' array and, if found, map the file to the
+    corresponding path in the matching item. If there is no match, then it checks if the file is included in the given 'paths'
+    array and maps the path to the given 'tmp_dir' (e.g. "/lib/systemd/system" becomes "<tmp_dir>/lib/systemd/system".) If
+    there no matches, the path is not changed. Once this mapping has completed the mocks invoke the corresponding original
+    function.
+
+    Matches are done using regular expressions; the regular expressions in 'paths' must create group 0 to indicate
+    the section of the path that needs to be mapped (i.e. use parenthesis around the section that needs to be mapped.)
+
+    The items in the given 'data_files' are copied to the 'tmp_dir'.
+
+    The add_*() methods insert new items int the list of mock objects. Items added by these methods take precedence over
+    the items provided to the __init__() method.
+    """
+    def __init__(self, tmp_dir, commands=None, paths=None, files=None, data_files=None):
+        # make a copy of the arrays passed as arguments since individual tests can modify them
+        self.tmp_dir = tmp_dir
+        self.commands = [] if commands is None else commands[:]
+        self.paths = [] if paths is None else paths[:]
+        self.files = [] if files is None else files[:]
+        self._data_files = data_files
+
+        # get references to the functions we'll mock so that we can call the original implementations
+        self._original_popen = subprocess.Popen
+        self._original_mkdir = fileutil.mkdir
+        self._original_path_exists = os.path.exists
+        self._original_open = open
+
+        self.patchers = [
+            patch_builtin("open", side_effect=self._mock_open),
+            patch("subprocess.Popen", side_effect=self._mock_popen),
+            patch("os.path.exists", side_effect=self._mock_path_exists),
+            patch("azurelinuxagent.common.utils.fileutil.mkdir", side_effect=self._mock_mkdir)
+        ]
+
+    def __enter__(self):
```

Comment:
Previously the contextmanager contract was implemented using the contextlib.contextmanager decorator; now it is explicitly implemented using __enter__ and __exit__.